File: benches/benchmark.rs
use criterion::{Criterion, criterion_group, criterion_main};
use php_hardened::sanitizers::html::{Flag, HtmlSanitizer};
use std::hint::black_box;

fn bench_htmlsanitizer(c: &mut Criterion) {
    fn payload(target_size: usize) -> String {
        let mut buf = String::from(
            r#"<!DOCTYPE html><html lang="ru"><head><meta charset="UTF - 8"><title>Benchmark HTML Purifier</title></head><body>"#,
        );
        while buf.len() < target_size {
            buf.push_str(r#"<div class="benchmark-item">Lorem ipsum dolor sit amet, <a href="javascript:alert(document.cookie)">consectetur</a> adipiscing elit. Quisque at.</div>"#);
        }
        buf.push_str(r#"</body></html>"#);
        buf
    }

    let payload10kb = payload(10 * 1024);
    let mut html_sanitizer = HtmlSanitizer::default();
    c.bench_function("html_sanitizer_10kb", |b| {
        b.iter(|| {
            let _ = black_box(html_sanitizer.clean(payload10kb.clone()).unwrap());
        })
    });

    c.bench_function("html_sanitizer_truncate_10k_to_5kb_in_ascii_mode", |b| {
        b.iter(|| {
            let _ = black_box(
                html_sanitizer
                    ._clean_and_truncate(payload10kb.clone(), 5 * 1024, &[Flag::Ascii], None)
                    .unwrap(),
            );
        })
    });
}

criterion_group!(benches, bench_htmlsanitizer);
criterion_main!(benches);

File: benches/benchmark.php
<?php
use Hardened\Sanitizers\HtmlSanitizer;

require __DIR__ . '/vendor/autoload.php';

class HtmlSanitizerBenchmark
{
    private $htmlSanitizer;
    private $ezyangHtmlPurifier;

    private $payload10kb;
    public function __construct() {
         $this->ezyangHtmlPurifier = new HTMLPurifier(HTMLPurifier_Config::createDefault());
         $this->htmlSanitizer = HtmlSanitizer::default();
         $this->payload10kb = $this->payload(10 * 1024);
    }

    public function payload(int $targetSize): String {
        $buffer = '<!DOCTYPE html><html lang="ru"><head><meta charset="UTF-8"><title>Benchmark HTML Purifier</title></head><body>';
        while (strlen($buffer) < $targetSize) {
            $buffer .= '<div class="benchmark-item">Lorem ipsum dolor sit amet, <a href="javascript:alert(document.cookie)">consectetur</a> adipiscing elit. Quisque at.</div>';
        }
        $buffer .= '</body></html>';
        return $buffer;
    }

    /**
     * @Revs(10000)
     */
    public function benchHtmlSanitizer10kb(): void
    {
         $this->htmlSanitizer->clean($this->payload10kb);
    }

    /**
     * @Revs(10000)
     */
    public function benchEzyangHtmlPurifier10kb(): void
    {
        $this->ezyangHtmlPurifier->purify($this->payload10kb);
    }

    /**
     * @Revs(10000)
     */
    public function benchTidy10kb(): void
    {
        $tidy = new tidy;
        $tidy->parseString($this->payload10kb, [], 'utf8');
        $tidy->cleanRepair();
    }
}
